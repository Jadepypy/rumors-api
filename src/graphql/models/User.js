import {
  GraphQLObjectType,
  GraphQLString,
  GraphQLInt,
  GraphQLNonNull,
} from 'graphql';
import crypto from 'crypto';
import {
  adjectives,
  names,
  towns,
  separators,
  decorators,
} from 'util/pseudonymDict';
import {
  accessories,
  faces,
  facialHairStyles,
  hairStyles,
  bustPoses,
} from 'util/openPeepsOptions';
import { sample, random } from 'lodash';

/**
 * Generates a pseudonym.
 */
export const generatePseudonym = () => {
  const [adj, name, place, separator, decorator] = [
    adjectives,
    names,
    towns,
    separators,
    decorators,
  ].map(ary => sample(ary));
  return decorator(separator({ adj, name, place }));
};

export const AvatarTypes = {
  OpenPeeps: 'OpenPeeps',
};

export const isBackendApp = appId =>
  appId !== 'WEBSITE' && appId !== 'DEVELOPMENT_FRONTEND';

/**
 * Generates data for open peeps avatar.
 */
export const generateOpenPeepsAvatar = () => {
  const accessory = random() ? sample(accessories) : 'None';
  const facialHair = random() ? sample(facialHairStyles) : 'None';
  const flip = !!random();
  const backgroundColorIndex = random(0, 1, true);

  const face = sample(faces);
  const hair = sample(hairStyles);
  const body = sample(bustPoses);

  return {
    accessory,
    body,
    face,
    hair,
    facialHair,
    backgroundColorIndex,
    flip,
  };
};

export const encodeAppId = appId =>
  crypto
    .createHash('md5')
    .update(appId)
    .digest('base64')
    .replace(/[+/]/g, '')
    .substr(0, 5);

export const sha256 = value =>
  crypto
    .createHash('sha256')
    .update(value)
    .digest('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '');

/**
 * @param {string} appUserId - user ID given by an backend app
 * @param {string} appId - app ID
 * @returns {string} the id used to index `user` in db
 */
export const convertAppUserIdToUserId = ({ appId, appUserId }) => {
  return `${encodeAppId(appId)}_${sha256(appUserId)}`;
};

/**
 * Field config helper for current user only field.
 * Resolves to null if root object is not current user. Otherwise, invoke provided resolver.
 *
 * @param {GraphQLScalarType | GraphQLObjectType} type
 * @param {function?} resolver - Use default resolver if not given.
 */
const currentUserOnlyField = (type, resolver) => ({
  type,
  description: 'Returns only for current user. Returns `null` otherwise.',
  resolve(user, arg, context, info) {
    if (user.id !== context.user.id) return null;

    return resolver ? resolver(user, arg, context, info) : user[info.fieldName];
  },
});

/**
 * Use Gravatar as avatar provider.
 */
const avatarResolver = (s = 80, d = 'identicon', r = 'g') => ({
  type: GraphQLString,
  description: 'return hash based on user email for gravatar url',
  resolve(user) {
    const GRAVATAR_URL = 'https://www.gravatar.com/avatar/';
    const hash = crypto
      .createHash('md5')
      .update((user.email || user.id).trim().toLocaleLowerCase())
      .digest('hex');
    const params = `?s=${s}&d=${d}&r=${r}`;
    return `${GRAVATAR_URL}${hash}${params}`;
  },
});

const User = new GraphQLObjectType({
  name: 'User',
  fields: () => ({
    id: { type: GraphQLString },
    email: currentUserOnlyField(GraphQLString),
    name: { type: GraphQLString },
    avatarUrl: avatarResolver(),
    //    avatarData: { type: GraphQLString },

    facebookId: currentUserOnlyField(GraphQLString),
    githubId: currentUserOnlyField(GraphQLString),
    twitterId: currentUserOnlyField(GraphQLString),
    repliedArticleCount: {
      type: GraphQLNonNull(GraphQLInt),
      description: 'Number of articles this user has replied to',
      resolve: (user, args, context) =>
        context.loaders.repliedArticleCountLoader
          .load(user.id)
          .then(num => num || 0),
    },
    votedArticleReplyCount: {
      type: GraphQLNonNull(GraphQLInt),
      description: 'Number of article replies this user has given feedbacks',
      resolve: (user, args, context) =>
        context.loaders.votedArticleReplyCountLoader
          .load(user.id)
          .then(num => num || 0),
    },

    level: {
      type: new GraphQLNonNull(GraphQLInt),
      async resolve(user, arg, context) {
        const { level } =
          (await context.loaders.userLevelLoader.load(user.id)) || {};
        return level || 0;
      },
    },
    points: {
      type: new GraphQLNonNull(
        new GraphQLObjectType({
          name: 'PointInfo',
          description:
            "Information of a user's point. Only available for current user.",
          fields: {
            total: {
              type: new GraphQLNonNull(GraphQLInt),
              description: 'Points earned by the current user',
            },
            currentLevel: {
              type: new GraphQLNonNull(GraphQLInt),
              description: 'Points required for current level',
            },
            nextLevel: {
              type: new GraphQLNonNull(GraphQLInt),
              description:
                'Points required for next level. null when there is no next level.',
            },
          },
        })
      ),
      resolve: async (user, arg, context) => {
        const { totalPoints, currentLevelPoints, nextLevelPoints } =
          (await context.loaders.userLevelLoader.load(user.id)) || {};

        return {
          total: totalPoints || 0,
          currentLevel: currentLevelPoints || 0,
          nextLevel: nextLevelPoints || 1,
        };
      },
    },
    createdAt: { type: GraphQLString },
    updatedAt: { type: GraphQLString },
    //    lastActiveAt: { type: GraphQLString },
  }),
});

export default User;

export const userFieldResolver = async (
  { userId, appId },
  args,
  { loaders, ...context }
) => {
  // If the root document is created by website users or if the userId is already converted to db userId, 
  // we can resolve user from userId.
  // 
  if (!isBackendApp(appId))
    return await loaders.docLoader.load({ index: 'users', id: userId });

  /*
  if (userId && userId.substr(0, 6) === `${encodeAppId(appId)}_`) {
    return await loaders.docLoader.load({ index: 'users', id: userId });
  }
  

  const user = await loaders.docLoader.load({ index: 'users', id: convertAppUserIdToUserId({ appId, appUserId: userId }) });
  if (user) return user;
  */

  // Todo: some unit tests are depending on this code block, need to clean up those tests and then remove the following lines.
  // If the user comes from the same client as the root document, return the user id.
  //
  if (context.appId === appId) return { id: userId };

  // If not, this client is not allowed to see user.
  //
  return null;
};
