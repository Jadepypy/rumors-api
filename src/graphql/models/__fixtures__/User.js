export default {
  '/users/doc/6LOqD_z5A8BwUr4gh1P2gw_2zFU3IIrSchTSl-vemod7CChMU': {
    name: 'test user 1',
    appUserId: 'userTest1',
    appId: 'TEST_BACKEND',
  },

  '/users/doc/userTest2': {
    name: 'test user 2',
    appId: 'WEBSITE',
  },
};
