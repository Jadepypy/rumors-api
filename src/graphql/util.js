import {
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLString,
  GraphQLInt,
  GraphQLList,
  GraphQLEnumType,
  GraphQLFloat,
  GraphQLNonNull,
  GraphQLID,
  GraphQLBoolean,
} from 'graphql';

import Connection from './interfaces/Connection';
import Edge from './interfaces/Edge';
import PageInfo from './interfaces/PageInfo';
import Highlights from './models/Highlights';
import client from 'util/client';
import { imageHash } from 'image-hash';

// https://www.graph.cool/docs/tutorials/designing-powerful-apis-with-graphql-query-parameters-aing7uech3
//
// Filtering args definition & parsing
//

/**
 * @param {string} typeName
 * @param {GraphQLScalarType} argType
 * @param {string} description
 * @returns {GraphQLInputObjectType}
 */
function getArithmeticExpressionType(typeName, argType, description) {
  return new GraphQLInputObjectType({
    name: typeName,
    description,
    fields: {
      LT: { type: argType },
      LTE: { type: argType },
      GT: { type: argType },
      GTE: { type: argType },
      EQ: { type: argType },
    },
  });
}

export const timeRangeInput = getArithmeticExpressionType(
  'TimeRangeInput',
  GraphQLString,
  'List only the entries that were created between the specific time range. ' +
    'The time range value is in elasticsearch date format (https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html)'
);
export const intRangeInput = getArithmeticExpressionType(
  'RangeInput',
  GraphQLInt,
  'List only the entries whose field match the criteria.'
);

/**
 * @param {object} arithmeticFilterObj - {LT, LTE, GT, GTE, EQ}, the structure returned by getArithmeticExpressionType
 * @returns {object} Elasticsearch range filter param
 * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html#range-query-field-params
 */
export function getRangeFieldParamFromArithmeticExpression(
  arithmeticFilterObj
) {
  // EQ overrides all other operators
  if (typeof arithmeticFilterObj.EQ !== 'undefined') {
    return {
      gte: arithmeticFilterObj.EQ,
      lte: arithmeticFilterObj.EQ,
    };
  }

  const conditionEntries = Object.entries(arithmeticFilterObj);

  if (conditionEntries.length === 0) throw new Error('Invalid Expression!');

  return Object.fromEntries(
    conditionEntries.map(([key, value]) => [key.toLowerCase(), value])
  );
}

export const moreLikeThisInput = new GraphQLInputObjectType({
  name: 'MoreLikeThisInput',
  description:
    'Parameters for Elasticsearch more_like_this query.\n' +
    'See: https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-mlt-query.html',
  fields: {
    like: {
      type: GraphQLString,
      description: 'The text string to search for.',
    },
    minimumShouldMatch: {
      type: GraphQLString,
      description:
        'more_like_this query\'s "minimum_should_match" query param.\n' +
        'See https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-minimum-should-match.html for possible values.',
    },
  },
});

export function createFilterType(typeName, args) {
  const filterType = new GraphQLInputObjectType({
    name: typeName,
    fields: () => ({
      ...args,
      // TODO: converting nested AND / OR to elasticsearch
      // AND: { type: new GraphQLList(filterType) },
      // OR: { type: new GraphQLList(filterType) },
    }),
  });
  return filterType;
}

//
// Sort args definition & parsing
//

const SortOrderEnum = new GraphQLEnumType({
  name: 'SortOrderEnum',
  values: {
    ASC: { value: 'asc' },
    DESC: { value: 'desc' },
  },
});

/**
 * @param {string} typeName
 * @param {Array<string|{name: string, description: string}>} filterableFields
 * @returns {GraphQLList<GarphQLInputObjectType>} sort input type for an field input argument.
 */
export function createSortType(typeName, filterableFields = []) {
  return new GraphQLList(
    new GraphQLInputObjectType({
      name: typeName,
      description:
        'An entry of orderBy argument. Specifies field name and the sort order. Only one field name is allowd per entry.',
      fields: filterableFields.reduce((fields, field) => {
        const fieldName = typeof field === 'string' ? field : field.name;
        const description =
          typeof field === 'string' ? undefined : field.description;

        return {
          ...fields,
          [fieldName]: { type: SortOrderEnum, description },
        };
      }, {}),
    })
  );
}

export const pagingArgs = {
  first: {
    type: GraphQLInt,
    description: 'Returns only first <first> results',
    defaultValue: 10,
  },
  after: {
    type: GraphQLString,
    description:
      'Specify a cursor, returns results after this cursor. cannot be used with "before".',
  },
  before: {
    type: GraphQLString,
    description:
      'Specify a cursor, returns results before this cursor. cannot be used with "after".',
  },
};

/**
 * @param {object[]} orderBy - sort input object type
 * @param {{[string]: (order: object) => object}} fieldFnMap - Defines one elasticsearch sort argument entry for a field
 * @returns {Array<{[string]: {order: string}}>} Elasticsearch sort argument in query body
 */
export function getSortArgs(orderBy, fieldFnMap = {}) {
  return orderBy
    .map(item => {
      const field = Object.keys(item)[0];
      const order = item[field];
      const defaultFieldFn = o => ({ [field]: { order: o } });

      return (fieldFnMap[field] || defaultFieldFn)(order);
    })
    .concat({ _id: { order: 'desc' } }); // enforce at least 1 sort order for pagination
}

// sort: [{fieldName: {order: 'desc'}}, {fieldName2: {order: 'desc'}}, ...]
// This utility function reverts the direction of each sort params.
//
function reverseSortArgs(sort) {
  if (!sort) return undefined;
  return sort.map(item => {
    const field = Object.keys(item)[0];
    const order = item[field].order === 'desc' ? 'asc' : 'desc';
    return {
      [field]: {
        ...item[field],
        order,
      },
    };
  });
}

// Export for custom resolveEdges() and resolveLastCursor()
//
export function getCursor(cursor) {
  return Buffer.from(JSON.stringify(cursor)).toString('base64');
}

export function getSearchAfterFromCursor(cursor) {
  if (!cursor) return undefined;
  return JSON.parse(Buffer.from(cursor, 'base64').toString('utf8'));
}

async function defaultResolveTotalCount({
  first, // eslint-disable-line no-unused-vars
  before, // eslint-disable-line no-unused-vars
  after, // eslint-disable-line no-unused-vars
  ...searchContext
}) {
  return (await client.count({
    ...searchContext,
    body: {
      ...searchContext.body,

      // totalCount cannot support these
      sort: undefined,
      track_scores: undefined,
    },
  })).body.count;
}

export async function defaultResolveEdges(
  { first, before, after, ...searchContext },
  args,
  { loaders }
) {
  if (before && after) {
    throw new Error('Use of before & after is prohibited.');
  }

  const nodes = await loaders.searchResultLoader.load({
    ...searchContext,
    body: {
      ...searchContext.body,
      size: first,
      search_after: getSearchAfterFromCursor(before || after),

      // if "before" is given, reverse the sort order and later reverse back
      //
      sort: before
        ? reverseSortArgs(searchContext.body.sort)
        : searchContext.body.sort,
      highlight: {
        order: 'score',
        fields: {
          text: {
            number_of_fragments: 1, // Return only 1 piece highlight text
            fragment_size: 200, // word count of highlighted fragment
            type: 'plain',
          },
          reference: {
            number_of_fragments: 1, // Return only 1 piece highlight text
            fragment_size: 200, // word count of highlighted fragment
            type: 'plain',
          },
        },
        pre_tags: ['<HIGHLIGHT>'],
        post_tags: ['</HIGHLIGHT>'],
      },
    },
  });

  if (before) {
    nodes.reverse();
  }

  return nodes.map(
    ({ _score: score, highlight, inner_hits, _cursor, ...node }) => ({
      node,
      cursor: getCursor(_cursor),
      score,
      highlight,
      inner_hits,
    })
  );
}

async function defaultResolveLastCursor(
  {
    first, // eslint-disable-line no-unused-vars
    before, // eslint-disable-line no-unused-vars
    after, // eslint-disable-line no-unused-vars
    ...searchContext
  },
  args,
  { loaders }
) {
  const lastNode = (await loaders.searchResultLoader.load({
    ...searchContext,
    body: {
      ...searchContext.body,
      sort: reverseSortArgs(searchContext.body.sort),
    },
    size: 1,
  }))[0];

  return lastNode && getCursor(lastNode._cursor);
}

async function defaultResolveFirstCursor(
  {
    first, // eslint-disable-line no-unused-vars
    before, // eslint-disable-line no-unused-vars
    after, // eslint-disable-line no-unused-vars
    ...searchContext
  },
  args,
  { loaders }
) {
  const firstNode = (await loaders.searchResultLoader.load({
    ...searchContext,
    size: 1,
  }))[0];

  return firstNode && getCursor(firstNode._cursor);
}

async function defaultResolveHighlights(edge) {
  const { highlight: { text, reference } = {}, inner_hits } = edge;

  const hyperlinks = inner_hits?.hyperlinks.hits.hits?.map(
    ({
      _source: { url },
      highlight: {
        'hyperlinks.title': title,
        'hyperlinks.summary': summary,
      } = {},
    }) => ({
      url,
      title: title ? title[0] : undefined,
      summary: summary ? summary[0] : undefined,
    })
  );

  // Elasticsearch highlight returns an array because it can be multiple fragments,
  // We directly returns first element(text, title, summary) here because we set number_of_fragments to 1.
  return {
    text: text ? text[0] : undefined,
    reference: reference ? reference[0] : undefined,
    hyperlinks,
  };
}

// All search
//
export function createConnectionType(
  typeName,
  nodeType,
  {
    // Default resolvers
    resolveTotalCount = defaultResolveTotalCount,
    resolveEdges = defaultResolveEdges,
    resolveLastCursor = defaultResolveLastCursor,
    resolveFirstCursor = defaultResolveFirstCursor,
    resolveHighlights = defaultResolveHighlights,
  } = {}
) {
  return new GraphQLObjectType({
    name: typeName,
    interfaces: [Connection],
    fields: () => ({
      totalCount: {
        type: new GraphQLNonNull(GraphQLInt),
        description:
          'The total count of the entire collection, regardless of "before", "after".',
        resolve: resolveTotalCount,
      },
      edges: {
        type: new GraphQLNonNull(
          new GraphQLList(
            new GraphQLNonNull(
              new GraphQLObjectType({
                name: `${typeName}Edge`,
                interfaces: [Edge],
                fields: {
                  node: { type: new GraphQLNonNull(nodeType) },
                  cursor: { type: new GraphQLNonNull(GraphQLString) },
                  score: { type: GraphQLFloat },
                  highlight: {
                    type: Highlights,
                    resolve: resolveHighlights,
                  },
                },
              })
            )
          )
        ),
        resolve: resolveEdges,
      },
      pageInfo: {
        type: new GraphQLNonNull(
          new GraphQLObjectType({
            name: `${typeName}PageInfo`,
            interfaces: [PageInfo],
            fields: {
              lastCursor: {
                type: GraphQLString,
                resolve: resolveLastCursor,
              },
              firstCursor: {
                type: GraphQLString,
                resolve: resolveFirstCursor,
              },
            },
          })
        ),
        resolve: params => params,
      },
    }),
  });
}

/**
 * @param {{status: T}[]} entriesWithStatus - list of objects with "status" field
 * @param {T[]} statuses - list of status to keep
 * @returns {Object[]}
 */
export function filterByStatuses(entriesWithStatus, statuses) {
  return entriesWithStatus.filter(({ status }) => statuses.includes(status));
}

/**
 * @param {Buffer} fileBuffer
 * @param {ArticleTypeEnum} type The article type
 * @returns {string} The hash for identifying if two files are similar
 */
export async function getMediaFileHash(fileBuffer, type) {
  let hash = '';
  if (type === 'IMAGE') {
    hash = await new Promise((resolve, reject) => {
      imageHash({ data: fileBuffer }, 16, true, (error, data) => {
        if (error) {
          console.error(error);
          reject(error);
        } else {
          resolve(data);
        }
      });
    });
  }

  return hash;
}

export const DEFAULT_ARTICLE_REPLY_STATUSES = ['NORMAL'];
export const DEFAULT_ARTICLE_CATEGORY_STATUSES = ['NORMAL'];
export const DEFAULT_REPLY_REQUEST_STATUSES = ['NORMAL'];
export const DEFAULT_ARTICLE_REPLY_FEEDBACK_STATUSES = ['NORMAL'];

/**
 * @param {string} pluralEntityName - the name to display on argument description
 * @returns {object} GraphQL args for common list filters
 */
export function createCommonListFilter(pluralEntityName) {
  return {
    appId: {
      type: GraphQLString,
      description: `Show only ${pluralEntityName} created by a specific app.`,
    },
    userId: {
      type: GraphQLString,
      description: `Show only ${pluralEntityName} created by the specific user.`,
    },
    createdAt: {
      type: timeRangeInput,
      description: `List only the ${pluralEntityName} that were created between the specific time range.`,
    },
    ids: {
      type: new GraphQLList(new GraphQLNonNull(GraphQLID)),
      description: `If given, only list out ${pluralEntityName} with specific IDs`,
    },
    selfOnly: {
      type: GraphQLBoolean,
      description: `Only list the ${pluralEntityName} created by the currently logged in user`,
    },
  };
}

/**
 * Attach (mutates) filterQueries with Elasticsearch query objects by args.filter in GraphQL resolver
 *
 * @param {Array<Object>} filterQueries - list of filter queries of Elasticsearch bool query
 * @param {object} filter - args.filter in resolver
 * @param {string} userId - userId for the currently logged in user
 * @param {string} appid - appId for the currently logged in user
 */
export function attachCommonListFilter(filterQueries, filter, userId, appId) {
  ['userId', 'appId'].forEach(field => {
    if (!filter[field]) return;
    filterQueries.push({ term: { [field]: filter[field] } });
  });

  if (filter.createdAt) {
    filterQueries.push({
      range: {
        createdAt: getRangeFieldParamFromArithmeticExpression(filter.createdAt),
      },
    });
  }

  if (filter.ids) {
    filterQueries.push({ ids: { values: filter.ids } });
  }

  if (filter.selfOnly) {
    if (!userId) throw new Error('selfOnly can be set only after log in');
    filterQueries.push({ term: { userId } }, { term: { appId } });
  }
}
