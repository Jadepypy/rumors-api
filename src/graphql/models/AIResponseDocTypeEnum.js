import { GraphQLEnumType } from 'graphql';

export default new GraphQLEnumType({
  name: 'AIResponseDocTypeEnum',
  values: {
    ARTICLE: { value: 'article' },
  },
});
