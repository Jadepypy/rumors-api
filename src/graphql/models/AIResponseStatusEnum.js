import { GraphQLEnumType } from 'graphql';

export default new GraphQLEnumType({
  name: 'AIResponseStatusEnum',
  values: {
    LOADING: { value: 'LOADING' },
    SUCCESS: { value: 'SUCCESS' },
    ERROR: { value: 'ERROR' },
  },
});
