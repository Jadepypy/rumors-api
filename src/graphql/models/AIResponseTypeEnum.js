import { GraphQLEnumType } from 'graphql';

export default new GraphQLEnumType({
  name: 'AIResponseTypeEnum',
  values: {
    AI_REPLY: {
      description:
        'The AI Response is an automated analysis / reply of an article.',
      value: 'AI_REPLY',
    },
    // TBA: speach-to-text result, OCR result, reply review response, etc
  },
});
