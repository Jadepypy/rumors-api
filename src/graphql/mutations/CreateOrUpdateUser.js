import { assertUser } from 'graphql/util';
import User, {
  generatePseudonym,
  generateOpenPeepsAvatar,
  AvatarTypes,
  isBackendApp,
  convertAppUserIdToUserId,
} from 'graphql/models/User';
import client, { processMeta } from 'util/client';

import rollbar from 'rollbarInstance';

/**
 * Index backend user if not existed, and record the last active time as now.
 *
 * @param {string} appUserId - user ID given by an backend app
 * @param {string} appId     - app ID
 *
 * @returns {user: User, isCreated: boolean}
 */
export async function createOrUpdateBackendUser({ appUserId, appId }) {
  assertUser({ appId, userId: appUserId });
  if (!isBackendApp(appId)) return { user: {}, isCreated: false };

  const now = new Date().toISOString();
  const dbUserId = convertAppUserIdToUserId({ appId, appUserId });

  const {
    body: { result, get: userFound },
  } = await client.update({
    index: 'users',
    type: 'doc',
    id: dbUserId,
    body: {
      doc: {
        lastActiveAt: now,
      },
      upsert: {
        name: generatePseudonym(),
        avatarType: AvatarTypes.OpenPeeps,
        avatarData: JSON.stringify(generateOpenPeepsAvatar()),
        appId,
        appUserId,
        createdAt: now,
        updatedAt: now,
        lastActiveAt: now,
      },
      _source: true,
    },
  });

  const isCreated = result === 'created';
  const user = processMeta({ ...userFound, _id: dbUserId });
  if (!isCreated && (user.appId !== appId || user.appUserId !== appUserId)) {
    const errorMessage = `collision found! ${
      user.appUserId
    } and ${appUserId} both hash to ${dbUserId}`;
    console.log(errorMessage);
    rollbar.error(`createBackendUserError: ${errorMessage}`);
  }
  return {
    user,
    isCreated,
  };
}

export default {
  description: 'Create or update a user for the given appId, appUserId pair',
  type: User,
  args: {},

  async resolve(rootValue, _, { appId, userId }) {
    const { user } = await createOrUpdateBackendUser({
      appId,
      appUserId: userId,
    });
    return user;
  },
};
