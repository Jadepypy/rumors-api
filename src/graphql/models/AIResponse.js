import {
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLString,
  GraphQLNonNull,
  GraphQLID,
  GraphQLInt,
} from 'graphql';

import { createConnectionType } from 'graphql/util';
import Node from '../interfaces/Node';
import AIResponseTypeEnum from './AIResponseTypeEnum';
import AIResponseStatusEnum from './AIResponseStatusEnum';
import User, { userFieldResolver } from './User';

const commonAiResponseFields = {
  id: { type: new GraphQLNonNull(GraphQLID) },

  docId: {
    description: 'The id for the document that this AI response is for.',
    type: GraphQLID,
  },

  type: {
    description: 'AI response type',
    type: new GraphQLNonNull(AIResponseTypeEnum),
  },

  user: {
    description: 'The user triggered this AI response',
    type: User,
    resolve: userFieldResolver,
  },

  status: {
    description: 'Processing status of AI',
    type: new GraphQLNonNull(AIResponseStatusEnum),
  },

  text: {
    description: 'AI response text. Populated after status becomes SUCCESS.',
    type: GraphQLString,
  },

  createdAt: { type: new GraphQLNonNull(GraphQLString) },
  updatedAt: { type: GraphQLString },
};

const AIResponse = new GraphQLInterfaceType({
  name: 'AIResponse',
  description: 'Denotes an AI processed response and its processing status.',
  interfaces: [Node],
  fields: commonAiResponseFields,
  resolveType(doc) {
    switch (doc.type) {
      case 'AI_REPLY':
        return AIReply;

      case 'TRANSCRIPT':
        return AITranscript;
    }
  },
});

export const AIReply = new GraphQLObjectType({
  name: 'AIReply',
  description: 'A ChatGPT reply for an article with no human fact-checks yet',
  interfaces: [Node, AIResponse],
  fields: {
    ...commonAiResponseFields,
    usage: {
      description:
        'The usage returned from OpenAI. Populated after status becomes SUCCESS.',
      type: new GraphQLObjectType({
        name: 'OpenAICompletionUsage',
        fields: {
          promptTokens: { type: new GraphQLNonNull(GraphQLInt) },
          completionTokens: { type: new GraphQLNonNull(GraphQLInt) },
          totalTokens: { type: new GraphQLNonNull(GraphQLInt) },
        },
      }),
    },
  },
});

export const AITranscript = new GraphQLObjectType({
  name: 'AITranscript',
  description: 'Transcript from OCR or speech-to-text AI models.',
  interfaces: [Node, AIResponse],
  fields: {
    ...commonAiResponseFields,
  },
});

export const AIResponseConnection = createConnectionType(
  'AIResponseConnection',
  AIResponse
);

export default AIResponse;
