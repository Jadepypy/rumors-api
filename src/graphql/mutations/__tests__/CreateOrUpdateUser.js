import { loadFixtures, unloadFixtures } from 'util/fixtures';
import client from 'util/client';
import MockDate from 'mockdate';
import fixtures from '../__fixtures__/CreateOrUpdateUser';
import rollbar from 'rollbarInstance';
import { getUserId } from 'graphql/models/User';
import { createOrUpdateUser } from '../CreateOrUpdateUser';

jest.mock('../../models/User', () => {
  const UserModel = jest.requireActual('../../models/User');
  return {
    ...UserModel,
    __esModule: true,
    generatePseudonym: jest
      .fn()
      .mockReturnValue('Friendly Neighborhood Spider Man'),
    generateOpenPeepsAvatar: jest.fn().mockReturnValue({ accessory: 'mask' }),
    getUserId: jest.spyOn(UserModel, 'getUserId'),
  };
});

describe('CreateOrUpdateUser', () => {
  beforeAll(() => loadFixtures(fixtures));

  afterAll(() => unloadFixtures(fixtures));

  it('creates backend user if not existed', async () => {
    MockDate.set(1602288000000);
    const userId = 'testUser2';
    const appId = 'TEST_BACKEND';

    const { user, isCreated } = await createOrUpdateUser({
      userId,
      appId,
    });

    expect(isCreated).toBe(true);
    expect(user).toMatchSnapshot();

    const id = getUserId({ userId, appId });

    const {
      body: { _source: source },
    } = await client.get({
      index: 'users',
      type: 'doc',
      id,
    });
    expect(source).toMatchSnapshot();
    expect(rollbar.error).not.toHaveBeenCalled();
    rollbar.error.mockClear();

    MockDate.reset();
    await client.delete({ index: 'users', type: 'doc', id });
  });

  it("updates backend users' last active time if user already existed", async () => {
    MockDate.set(1602291600000);

    const userId = 'testUser1';
    const appId = 'TEST_BACKEND';

    const { user, isCreated } = await createOrUpdateUser({
      userId,
      appId,
    });

    expect(isCreated).toBe(false);
    expect(user).toMatchSnapshot();

    const id = getUserId({ userId, appId });
    const {
      body: { _source: source },
    } = await client.get({
      index: 'users',
      type: 'doc',
      id,
    });
    expect(source).toMatchSnapshot();
    expect(rollbar.error).not.toHaveBeenCalled();
    rollbar.error.mockClear();
  });

  it('logs error if collision occurs', async () => {
    MockDate.set(1602291600000);

    const userId = 'testUser3';
    const appId = 'TEST_BACKEND';
    const id = getUserId({ userId: 'testUser1', appId });

    getUserId.mockReturnValueOnce(id);
    const { user, isCreated } = await createOrUpdateUser({
      userId,
      appId,
    });

    expect(isCreated).toBe(false);
    expect(user).toMatchSnapshot();

    const {
      body: { _source: source },
    } = await client.get({
      index: 'users',
      type: 'doc',
      id,
    });
    expect(source).toMatchSnapshot();
    expect(rollbar.error.mock.calls).toMatchSnapshot();
    rollbar.error.mockClear();
  });
});
