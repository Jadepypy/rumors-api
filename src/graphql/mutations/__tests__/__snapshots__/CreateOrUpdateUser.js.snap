// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateOrUpdateUser creates backend user if not existed 1`] = `
Object {
  "CreateOrUpdateUser": Object {
    "createdAt": "2020-10-10T00:00:00.000Z",
    "id": "6LOqD_3gpe4ZVaxRvemf7KNTfm6y3WNBu1hbs-5MRdSWiWVss",
    "name": "Friendly Neighborhood Spider Man",
    "updatedAt": "2020-10-10T00:00:00.000Z",
  },
}
`;

exports[`CreateOrUpdateUser creates backend user if not existed 2`] = `
Object {
  "appId": "TEST_BACKEND",
  "appUserId": "testUser2",
  "avatarData": "{\\"accessory\\":\\"mask\\"}",
  "avatarType": "OpenPeeps",
  "createdAt": "2020-10-10T00:00:00.000Z",
  "lastActiveAt": "2020-10-10T00:00:00.000Z",
  "name": "Friendly Neighborhood Spider Man",
  "updatedAt": "2020-10-10T00:00:00.000Z",
}
`;

exports[`CreateOrUpdateUser logs error if collision occurs 1`] = `
Object {
  "CreateOrUpdateUser": Object {
    "id": "6LOqD_QabTT7XXTrsz7ybEa5PLfc0GfXlV578HYPhODPfSWc8",
    "name": "test user 1",
  },
}
`;

exports[`CreateOrUpdateUser logs error if collision occurs 2`] = `
Object {
  "appId": "TEST_BACKEND",
  "appUserId": "testUser1",
  "lastActiveAt": "2020-10-10T01:00:00.000Z",
  "name": "test user 1",
}
`;

exports[`CreateOrUpdateUser logs error if collision occurs 3`] = `
Array [
  Array [
    "createBackendUserError: collision found! testUser1 and testUser2 both hash to 6LOqD_QabTT7XXTrsz7ybEa5PLfc0GfXlV578HYPhODPfSWc8",
  ],
]
`;

exports[`CreateOrUpdateUser updates backend users' last active time if user already existed 1`] = `
Object {
  "CreateOrUpdateUser": Object {
    "createdAt": null,
    "id": "6LOqD_QabTT7XXTrsz7ybEa5PLfc0GfXlV578HYPhODPfSWc8",
    "name": "test user 1",
    "updatedAt": null,
  },
}
`;

exports[`CreateOrUpdateUser updates backend users' last active time if user already existed 2`] = `
Object {
  "appId": "TEST_BACKEND",
  "appUserId": "testUser1",
  "lastActiveAt": "2020-10-10T01:00:00.000Z",
  "name": "test user 1",
}
`;
